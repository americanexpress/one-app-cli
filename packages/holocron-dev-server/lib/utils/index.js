/*
 * Copyright 2021 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

export * from './helpers';
export * from './paths';
export * from './publish';
export * from './virtual-file-system';
export * from './watcher';

export {
  loadModuleLanguagePack,
  loadModuleLanguagePacks,
  writeModuleLanguagePacksToVolume,
  addModuleLanguagePackToVolume,
  removeModuleLanguagePackFromVolume,
  loadLanguagePacks,
} from './language-packs';

export {
  loadRemoteModuleMap,
  createLocalModuleMap,
  createUnifiedModuleMap,
  createModuleMap,
} from './module-map';

export {
  renderDocument,
  createInitialState,
  getEntryScriptsForExternals,
  getWebpackScriptsForLocalModules,
} from './rendering';

export { loadStatics, loadOneAppStaticsFromDocker, addStaticsDirToGitIgnore } from './statics';
