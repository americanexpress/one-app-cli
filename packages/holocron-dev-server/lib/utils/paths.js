/*
 * Copyright 2021 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import path from 'path';

import { modulesBundleName } from '../constants';

export const STATIC_DIR = 'static';
export const ONE_APP_DIR = 'app';
export const MODULES_DIR = 'modules';
export const EXTERNAL_DIR = 'vendor';
export const TEMP_DIR = '.temp';

// file paths that are primarily used to place all build assets

export function getContextPath(filePath = '') {
  return path.join(process.cwd(), filePath);
}

export function getStaticPath(filePath = '', context) {
  return path.join(context || getContextPath(), STATIC_DIR, filePath);
}

export function getModulesPath(filePath = '', context) {
  return path.join(context || getContextPath(), STATIC_DIR, MODULES_DIR, filePath);
}

export function getOneAppPath(filePath = '', context) {
  return path.join(context || getContextPath(), STATIC_DIR, ONE_APP_DIR, filePath);
}

export function getVendorsPath(filePath = '', context) {
  return path.join(context || getContextPath(), STATIC_DIR, EXTERNAL_DIR, filePath);
}

export function getTempPath(filePath = '', context) {
  return path.join(context || getContextPath(), STATIC_DIR, TEMP_DIR, filePath);
}

export function getMockDirectoryForModule(modulePath = getContextPath()) {
  return path.join(modulePath, 'mock');
}

export function getScenarioPathForModule(modulePath = getContextPath()) {
  return path.join(modulePath, 'mock', 'scenarios.js');
}

export function getLocalesPathForModule(modulePath = getContextPath()) {
  return path.join(modulePath, 'locale');
}

// URLs relatively formatted to the development server

export function combineUrlFragments(...args) {
  return [].concat(args).join('/');
}

export function joinUrlFragments(...args) {
  return []
    .concat(args)
    .map((pathName) => pathName.replace(/^\//, '').replace(/\/$/, ''))
    .filter((pathName) => !!pathName)
    .join('/');
}

export function getPublicUrl(filePath = '') {
  return `/${combineUrlFragments(...[STATIC_DIR].concat(filePath || []))}`;
}

export function getPublicModulesUrl(moduleName) {
  return `/${combineUrlFragments(STATIC_DIR, MODULES_DIR, moduleName || [])}`;
}

export function getPublicAppUrl(appPath) {
  return `/${combineUrlFragments(STATIC_DIR, ONE_APP_DIR, appPath || [])}`;
}

export function getPublicVendorsUrl(resource) {
  return `/${combineUrlFragments(STATIC_DIR, EXTERNAL_DIR, resource || [])}`;
}

export function createModuleScriptUrl(moduleName, bundleType, version) {
  return joinUrlFragments(
    version ? [moduleName, version].join('/') : moduleName,
    `${bundleType ? [moduleName, bundleType].join('.') : moduleName}.js`
  );
}

// file names

export function getReportFilename(name = modulesBundleName) {
  return `${name}-report.html`;
}
