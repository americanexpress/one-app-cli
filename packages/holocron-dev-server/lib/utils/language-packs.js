/*
 * Copyright 2021 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import path from 'path';
import jsonParse from 'json-parse-context';

import { vol, ufs } from './virtual-file-system';
import { getModulesPath, getLocalesPathForModule } from './paths';
import { logModuleLanguagePacksLoaded } from './logs';
import { createLanguagePackWatcher } from './watcher';

export const copyFileName = 'copy.json';
export const linksFileName = 'integration.json';
export const linksPathName = `links/${linksFileName}`;
export const localePathName = 'locale';

export function extractLanguageDataFromLocale(languageDataPath) {
  const fileStats = ufs.statSync(languageDataPath);
  if (fileStats.isDirectory()) {
    const copy = extractLanguageDataFromLocale(`${languageDataPath}/${copyFileName}`);
    const links = extractLanguageDataFromLocale(`${languageDataPath}/${linksPathName}`) || {};
    return {
      ...copy,
      links,
    };
  }
  if (fileStats.isFile()) {
    const json = ufs.readFileSync(languageDataPath).toString();
    return jsonParse(json);
  }
  return null;
}

export function loadModuleLanguagePacks({ modulePath, localePath = localePathName }) {
  const languagePacksPath = path.join(modulePath, localePath);
  if (ufs.existsSync(languagePacksPath)) {
    return ufs
      .readdirSync(languagePacksPath)
      .map((locale) => [
        locale.replace(/(\..*)$/, '').toLowerCase(),
        extractLanguageDataFromLocale(path.join(languagePacksPath, locale)),
      ]);
  }
  return [];
}

export function writeModuleLanguagePacksToVolume({
  modulePath,
  moduleName,
  localePath = localePathName,
}) {
  const languagePacks = loadModuleLanguagePacks({ modulePath, localePath });
  const locales = languagePacks.reduce(
    (map, [locale, langPack]) => ({
      ...map,
      [[locale, `${moduleName}.json`].join('/')]: JSON.stringify(langPack),
    }),
    {}
  );
  vol.fromJSON(locales, getModulesPath(moduleName));
  return languagePacks.map(([locale]) => locale);
}

export function addModuleLanguagePackToVolume({ filePath, moduleName, locale }) {
  const localeSymbol = locale.replace(/(\..*)$/, '').toLowerCase();
  const langPack = extractLanguageDataFromLocale(filePath);
  const bundlePath = getModulesPath([moduleName, localeSymbol, `${moduleName}.json`].join('/'));
  vol.writeFileSync(bundlePath, JSON.stringify(langPack));
}

export function removeModuleLanguagePackFromVolume({ moduleName, locale }) {
  const localeFilePath = getModulesPath(
    [moduleName, locale.replace(/(\..*)$/, '').toLowerCase(), `${moduleName}.json`].join('/')
  );
  vol.unlinkSync(localeFilePath);
  vol.rmdirSync(localeFilePath.replace(`/${moduleName}.json`, ''));
}

export function loadLanguagePacks({ modules = [] } = {}) {
  const modulePaths = modules
    .filter(({ modulePath }) => ufs.existsSync(getLocalesPathForModule(modulePath)))
    .map(({ modulePath }) => modulePath);

  if (modulePaths.length > 0) {
    modulePaths.forEach((modulePath) => {
      const [moduleName] = modulePath.split('/').reverse();
      const languagePacks = writeModuleLanguagePacksToVolume({ modulePath, moduleName });
      logModuleLanguagePacksLoaded({ moduleName, languagePacks });
    });
    return createLanguagePackWatcher(
      { modules },
      {
        add: addModuleLanguagePackToVolume,
        change: addModuleLanguagePackToVolume,
        remove: removeModuleLanguagePackFromVolume,
      }
    );
  }
  return Promise.resolve();
}
