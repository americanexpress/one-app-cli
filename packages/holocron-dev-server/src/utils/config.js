/*
 * Copyright 2021 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import path from 'node:path';
import readPkgUp from 'read-pkg-up';

import { defaultLogLevel, errorReportingUrlFragment, oneAppDockerImageName } from '../constants';
import { createModuleScriptUrl, getContextPath, getPublicModulesUrl } from './paths';

export const ONE_AMEX_CONFIG_KEY = 'one-amex';
export const BUNDLER_CONFIG_KEY = 'bundler';
export const RUNNER_CONFIG_KEY = 'runner';
export const HMR_CONFIG_KEY = 'hmr';

async function getPackageJsonConfig(modulePath = getContextPath()) {
  const {
    packageJson: {
      [ONE_AMEX_CONFIG_KEY]: {
        [BUNDLER_CONFIG_KEY]: bundler,
        [RUNNER_CONFIG_KEY]: runner,
        [HMR_CONFIG_KEY]: hmr,
      } = {},
      name: moduleName,
      version: moduleVersion,
    },
  } = await readPkgUp({ cwd: modulePath });
  return {
    hmr,
    runner,
    bundler,
    moduleName,
    moduleVersion,
    modulePath,
  };
}

export function extractRunnerOptions({
  modules = [],
  rootModuleName,
  moduleMapUrl,
  dockerImage,
  envVars,
} = {}) {
  return {
    modules: modules.map((relativeModulePath) => path.resolve(getContextPath(), relativeModulePath)
    ),
    remoteModuleMapUrl: moduleMapUrl,
    environmentVariables: envVars,
    rootModuleName,
    dockerImage,
  };
}

export function extractBundlerOptions({
  providedExternals = [],
  requiredExternals = [],
  performanceBudget,
  webpackConfigPath,
  webpackClientConfigPath,
  purgecss,
} = {}) {
  return {
    providedExternals,
    requiredExternals,
    webpackConfigPath: webpackClientConfigPath || webpackConfigPath,
    performanceBudget,
    purgecss,
  };
}

export function extractHmrOptions({
  clientConfig = {},
  sourceMap,
  openWhenReady = false,
  logLevel,
  port,
} = {}) {
  return {
    clientConfig: {
      errorReportingUrl: errorReportingUrlFragment,
      ...clientConfig,
    },
    sourceMap,
    openWhenReady,
    logLevel,
    port,
  };
}

export function createModulesConfig(context) {
  return Promise.all(context.modules.map(getPackageJsonConfig))
    .then((modules) => modules.map(({
      bundler, runner, hmr, ...moduleConfig
    }) => {
      const {
        providedExternals, requiredExternals, environmentVariables,
      } = {
        ...extractBundlerOptions(bundler),
        ...extractRunnerOptions(runner),
        ...extractHmrOptions(hmr),
      };
      const { moduleName } = moduleConfig;
      return {
        ...moduleConfig,
        moduleName,
        providedExternals,
        requiredExternals,
        environmentVariables,
        rootModule: moduleName === context.rootModuleName,
        // add the local url path for the module
        src: getPublicModulesUrl(createModuleScriptUrl(moduleName)),
      };
    }));
}

export function createConfigurationContext({
  bundler,
  runner,
  hmr,
  moduleName,
  modulePath,
  moduleVersion,
}) {
  const {
    modules,
    rootModuleName = moduleName,
    dockerImage = oneAppDockerImageName,
    logLevel = defaultLogLevel,
    port = 4000,
    remoteModuleMapUrl,
    environmentVariables,
    openWhenReady,
    clientConfig,
    sourceMap,
    webpackConfigPath,
    performanceBudget,
    purgecss,
  } = {
    ...extractBundlerOptions(bundler),
    ...extractRunnerOptions(runner),
    ...extractHmrOptions(hmr),
  };

  if (!modules.find((pathName) => pathName === modulePath)) {
    modules.unshift(modulePath);
  }

  return {
    moduleName,
    modulePath,
    moduleVersion,
    modules,
    logLevel,
    port,
    dockerImage,
    rootModuleName,
    remoteModuleMapUrl,
    environmentVariables,
    openWhenReady,
    clientConfig,
    sourceMap,
    webpackConfigPath,
    performanceBudget,
    purgecss,
  };
}

export async function createConfig() {
  const entryConfig = await getPackageJsonConfig();
  const context = createConfigurationContext(entryConfig);
  const modules = await createModulesConfig(context);

  const serverAddress = `http://localhost:${context.port}/`;

  return {
    ...context,
    modules,
    serverAddress,
  };
}
