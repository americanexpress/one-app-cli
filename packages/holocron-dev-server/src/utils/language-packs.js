/*
 * Copyright 2021 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import path from 'path';
import jsonParse from 'json-parse-context';

import { volume, ufs } from './virtual-file-system';
import { getModulesPath, getLocalesPathForModule } from './paths';
import { logModuleLanguagePacksLoaded } from './logs';
import { createLanguagePackWatcher } from './watcher';

export const localePathName = 'locale';
export const languagePackFileNames = ['copy', 'integration'];

export function readJsonFile(filePath) {
  const json = ufs.readFileSync(filePath).toString();
  return jsonParse(json);
}

export function loadModuleLanguagePack({
  moduleName, modulePath, localePath = localePathName, locale,
}) {
  let languagePackPath = path.join(modulePath, localePath, locale);

  if (!ufs.existsSync(languagePackPath)) {
    languagePackPath = path.join(modulePath, localePath, `${locale}.json`);
  }

  if (ufs.existsSync(languagePackPath)) {
    const fileStats = ufs.statSync(languagePackPath);
    if (fileStats.isDirectory()) {
      const languagePack = {};
      const scan = [
        ...languagePackFileNames,
        locale,
        moduleName,
      ];
      scan
        .map((fileName) => `${fileName}.json`)
        .map((fileName) => path.join(languagePackPath, fileName))
        .filter((filePath) => ufs.existsSync(filePath))
        .forEach((filePath) => {
          Object.assign(languagePack, readJsonFile(filePath));
        }
        );
      const directories = ufs.readdirSync(languagePackPath, { withFileTypes: true })
        .filter((files) => files.isDirectory()).map((files) => files.name);
      if (directories.length > 0) {
        directories.forEach((directoryName) => {
          scan
            .map((fileName) => path.join(languagePackPath, directoryName, `${fileName}.json`))
            .filter((directoryFilePath) => ufs.existsSync(directoryFilePath))
            .forEach((directoryFilePath) => {
              languagePack[directoryName] = readJsonFile(directoryFilePath);
            });
        }
        );
      }
      return languagePack;
    }
    if (fileStats.isFile()) {
      return readJsonFile(languagePackPath);
    }
  }

  return null;
}

export function loadModuleLanguagePacks({ moduleName, modulePath, localePath = localePathName }) {
  const languagePacksPath = path.join(modulePath, localePath);
  if (!ufs.existsSync(languagePacksPath)) return [];

  return ufs
    .readdirSync(languagePacksPath)
    .map((locale) => locale.replace(/(\..*)$/, '').toLowerCase())
    .map((locale) => [
      locale,
      loadModuleLanguagePack({
        moduleName, modulePath, localePath, locale,
      }),
    ])
    .filter(([, languagePack]) => !!languagePack);
}

export function writeModuleLanguagePacksToVolume({
  modulePath,
  moduleName,
  localePath = localePathName,
}) {
  const languagePacks = loadModuleLanguagePacks({ moduleName, modulePath, localePath });
  const locales = languagePacks.reduce(
    (map, [locale, langPack]) => ({
      ...map,
      [[locale, `${moduleName}.json`].join('/')]: JSON.stringify(langPack),
    }),
    {}
  );
  volume.fromJSON(locales, getModulesPath(moduleName));
  return languagePacks.map(([locale]) => locale);
}

export function addModuleLanguagePackToVolume({ filePath, moduleName, locale }) {
  const localeSymbol = locale.replace(/(\..*)$/, '').toLowerCase();
  const modulePath = filePath.split('/').reduce((pathName, next) => (pathName.endsWith(moduleName) ? pathName : [pathName, next].join('/')), '');
  const langPack = loadModuleLanguagePack({
    moduleName, modulePath, locale: localeSymbol,
  });
  const bundlePath = getModulesPath([moduleName, localeSymbol, `${moduleName}.json`].join('/'));
  volume.writeFileSync(bundlePath, JSON.stringify(langPack));
}

export function removeModuleLanguagePackFromVolume({ moduleName, locale }) {
  const localeFilePath = getModulesPath(
    [moduleName, locale.replace(/(\..*)$/, '').toLowerCase(), `${moduleName}.json`].join('/')
  );
  volume.unlinkSync(localeFilePath);
  volume.rmdirSync(localeFilePath.replace(`/${moduleName}.json`, ''));
}

export function loadLanguagePacks({ modules = [] } = {}) {
  const modulePaths = modules
    .filter(({ modulePath }) => ufs.existsSync(getLocalesPathForModule(modulePath)))
    .map(({ modulePath }) => modulePath);

  if (modulePaths.length > 0) {
    modulePaths.forEach((modulePath) => {
      const [moduleName] = modulePath.split(path.sep).reverse();
      const languagePacks = writeModuleLanguagePacksToVolume({ modulePath, moduleName });
      logModuleLanguagePacksLoaded({ moduleName, languagePacks });
    });
    return createLanguagePackWatcher(
      { modules },
      {
        add: addModuleLanguagePackToVolume,
        change: addModuleLanguagePackToVolume,
        remove: removeModuleLanguagePackFromVolume,
      }
    );
  }
  return Promise.resolve();
}
