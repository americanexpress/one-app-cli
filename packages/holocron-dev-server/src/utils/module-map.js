/*
 * Copyright 2021 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import fetch from 'cross-fetch';
import { createTimeoutFetch } from '@americanexpress/fetch-enhancers';

import { createModuleScriptUrl, getPublicModulesUrl, getStaticPath } from './paths';
import {
  debug,
  logLocalModulesLoaded,
  logRemoteModulesLoaded,
  errorOnRemoteModuleMapResponse,
  errorOnRemoteModuleMapFetching,
} from './logs';
import { bundleType as defaultBundleType } from '../constants';
import { volume } from './virtual-file-system';

export async function loadRemoteModuleMap(remoteModuleMapUrl) {
  const fetcher = createTimeoutFetch(6e3)(fetch);
  if (remoteModuleMapUrl) {
    try {
      const response = await fetcher(remoteModuleMapUrl);
      if (response.ok) {
        return response.json();
      }
      errorOnRemoteModuleMapResponse();
    } catch (e) {
      errorOnRemoteModuleMapFetching(e);
    }
  }

  return {
    modules: {},
  };
}

export function createLocalModuleMap(modules = [], bundleType) {
  return {
    modules: modules.reduce((map, { moduleName }) => {
      const baseUrl = getPublicModulesUrl(moduleName);
      return {
        ...map,
        [moduleName]: {
          baseUrl: `${baseUrl}/`,
          [bundleType || defaultBundleType]: {
            // TODO: only works with "browser", "node" for path name..
            // not "legacyBrowser" - "..legacy.browser.js"
            url: getPublicModulesUrl(createModuleScriptUrl(moduleName, bundleType)),
          },
        },
      };
    }, {}),
  };
}

export function createUnifiedModuleMap({ localModuleMap, remoteModuleMap, bundleType }) {
  const localModuleKeys = Object.keys(localModuleMap.modules);
  const remoteModules = Object.keys(remoteModuleMap.modules)
    .filter((moduleName) => !localModuleKeys.includes(moduleName))
    .map((moduleName) => ({ moduleName }));
  const localizedRemoteModuleMap = createLocalModuleMap(remoteModules, bundleType);
  return {
    modules: {
      ...localizedRemoteModuleMap.modules,
      ...localModuleMap.modules,
    },
  };
}

export async function createModuleMap({
  modules,
  remoteModuleMapUrl,
  bundleType = defaultBundleType,
}) {
  const remoteModuleMap = await loadRemoteModuleMap(remoteModuleMapUrl);
  const localModuleMap = createLocalModuleMap(modules);
  const moduleMap = createUnifiedModuleMap({
    localModuleMap,
    remoteModuleMap,
    bundleType,
  });

  const localModuleNames = Object.keys(localModuleMap.modules);
  const remoteModuleNames = Object.keys(remoteModuleMap.modules);
  // TODO: log module count, stats
  logLocalModulesLoaded(localModuleNames);
  logRemoteModulesLoaded(remoteModuleNames, localModuleNames);

  // add it to vfs to be be served by dev middleware
  volume.fromJSON(
    {
      'module-map.json': JSON.stringify(moduleMap, null, 2),
      'local-module-map.json': JSON.stringify(localModuleMap, null, 2),
      'remote-module-map.json': JSON.stringify(remoteModuleMap, null, 2),
    },
    getStaticPath()
  );

  debug('module map %o', moduleMap);

  return {
    remoteModuleMap,
    localModuleMap,
    moduleMap,
  };
}
