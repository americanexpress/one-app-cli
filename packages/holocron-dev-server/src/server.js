/*
 * Copyright 2021 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import express from 'express';

import createRenderingMiddleware from './middleware/html';
import requestAcceptedMiddleware from './middleware/request-accepted';
import createMocksMiddleware from './middleware/parrot-scenarios';
import createModulesProxyRelayMiddleware from './middleware/proxy-relay';
import loadWebpackMiddleware from './middleware/webpack';

import { createModuleMap } from './utils/module-map';
import { getPublicUrl, getStaticPath } from './utils/paths';
import { isDevelopment, openBrowser } from './utils/helpers';
import { loadLanguagePacks } from './utils/language-packs';
import { loadStatics } from './utils/statics';
import {
  setLogLevel,
  logHotReloadReady,
  logServerUrl,
  logModuleBundlerAnalyzerUrl,
  logServerStart,
  logError,
} from './utils/logs';
import { errorReportingUrlFragment } from './constants';

export function onLaunch({
  openWhenReady, serverAddress, port,
}) {
  return (error) => {
    if (error) throw error;
    logServerUrl(serverAddress, port);
    logHotReloadReady();
    logModuleBundlerAnalyzerUrl(serverAddress);
    if (openWhenReady) openBrowser(serverAddress);
  };
}

export default async function holocronDevServer({
  port = 4000,
  serverAddress = `http://localhost:${port}/`,
  logLevel,
  clientConfig,
  openWhenReady,
  modules,
  rootModuleName,
  environmentVariables,
  remoteModuleMapUrl,
  dockerImage,
  webpackConfigPath,
}) {
  if (!isDevelopment()) {
    logError(
      'Please ensure you are running this in development environment. Check that NODE_ENV="development"'
    );
    // Since this is a CLI tool and only enabled for production build
    // eslint-disable-next-line unicorn/no-process-exit
    process.exit(1);
  }
  setLogLevel(logLevel);

  logServerStart({ rootModuleName });

  await loadStatics({ dockerImage });
  await loadLanguagePacks({ modules });

  const { moduleMap, localModuleMap, remoteModuleMap } = await createModuleMap({
    modules,
    remoteModuleMapUrl,
  });
  // TODO: look are re-using one-app-dev-cdn and one-add-dev-proxy
  const proxyRelayMiddleware = createModulesProxyRelayMiddleware({
    moduleMap,
    localModuleMap,
    remoteModuleMap,
  });

  const mockMiddleware = await createMocksMiddleware({
    modules,
    serverAddress,
  });

  const [devMiddleware, hotMiddleware] = await loadWebpackMiddleware({
    modules,
    environmentVariables,
    webpackConfigPath,
  });

  const renderMiddleware = createRenderingMiddleware({
    rootModuleName,
    moduleMap,
    modules,
    clientConfig,
  });

  const app = express();

  app
    .post(errorReportingUrlFragment, requestAcceptedMiddleware)
    .use(proxyRelayMiddleware)
    .use(devMiddleware)
    .use(hotMiddleware)
    .use(getPublicUrl(), express.static(getStaticPath()))
    .use(mockMiddleware)
    .get('*', renderMiddleware);

  app.start = () => app.listen(
    port,
    onLaunch({
      port,
      serverAddress,
      openWhenReady,
    })
  );

  return app;
}
