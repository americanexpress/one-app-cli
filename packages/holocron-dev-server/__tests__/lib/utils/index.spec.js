/*
 * Copyright 2021 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import * as utils from '../../../src/utils';

test('exports utils API functions', () => {
  expect(Object.keys(utils)).toEqual([
    'loadModuleLanguagePack',
    'loadModuleLanguagePacks',
    'writeModuleLanguagePacksToVolume',
    'addModuleLanguagePackToVolume',
    'removeModuleLanguagePackFromVolume',
    'loadLanguagePacks',
    'loadRemoteModuleMap',
    'createLocalModuleMap',
    'createUnifiedModuleMap',
    'createModuleMap',
    'renderDocument',
    'createInitialState',
    'getEntryScriptsForExternals',
    'getWebpackScriptsForLocalModules',
    'loadStatics',
    'loadOneAppStaticsFromDocker',
    'addStaticsDirToGitIgnore',
    'isDevelopment',
    'getLocalRootModule',
    'getModuleFromFilePath',
    'getModuleInfoFromLocalePath',
    'openBrowser',
    'getContextPath',
    'getStaticPath',
    'getModulesPath',
    'getOneAppPath',
    'getVendorsPath',
    'getTempPath',
    'getMockDirectoryForModule',
    'getScenarioPathForModule',
    'getLocalesPathForModule',
    'combineUrlFragments',
    'joinUrlFragments',
    'getPublicUrl',
    'getPublicModulesUrl',
    'getPublicAppUrl',
    'getPublicVendorsUrl',
    'createModuleScriptUrl',
    'getReportFilename',
    'STATIC_DIR',
    'ONE_APP_DIR',
    'MODULES_DIR',
    'EXTERNAL_DIR',
    'TEMP_DIR',
    'clearPublisher',
    'setPublisher',
    'publish',
    'volume',
    'vfs',
    'ufs',
    'watchFiles',
    'createLanguagePackWatchEventHandler',
    'createLanguagePackWatcher',
  ]);
});
