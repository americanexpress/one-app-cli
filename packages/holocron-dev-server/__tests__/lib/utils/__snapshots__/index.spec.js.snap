// Jest Snapshot v1, https://goo.gl/fbAQLP
// eslint-disable-next-line jest/no-large-snapshots
exports[`exports utils API functions 1`] = `
Array [
  "extractLanguageDataFromLocale",
  "loadModuleLanguagePacks",
  "writeModuleLanguagePacksToVolume",
  "addModuleLanguagePackToVolume",
  "removeModuleLanguagePackFromVolume",
  "loadLanguagePacks",
  "loadRemoteModuleMap",
  "createLocalModuleMap",
  "createUnifiedModuleMap",
  "createModuleMap",
  "renderDocument",
  "createInitialState",
  "getEntryScriptsForExternals",
  "getWebpackScriptsForLocalModules",
  "loadStatics",
  "loadOneAppStaticsFromDocker",
  "addStaticsDirToGitIgnore",
  "isDevelopment",
  "getLocalRootModule",
  "getModuleFromFilePath",
  "getModuleInfoFromLocalePath",
  "openBrowser",
  "getContextPath",
  "getStaticPath",
  "getModulesPath",
  "getOneAppPath",
  "getVendorsPath",
  "getTempPath",
  "getMockDirectoryForModule",
  "getScenarioPathForModule",
  "getLocalesPathForModule",
  "combineUrlFragments",
  "joinUrlFragments",
  "getPublicUrl",
  "getPublicModulesUrl",
  "getPublicAppUrl",
  "getPublicVendorsUrl",
  "createModuleScriptUrl",
  "getReportFilename",
  "STATIC_DIR",
  "ONE_APP_DIR",
  "MODULES_DIR",
  "EXTERNAL_DIR",
  "TEMP_DIR",
  "clearPublisher",
  "setPublisher",
  "publish",
  "vol",
  "vfs",
  "ufs",
  "watchFiles",
  "createLanguagePackWatchEventHandler",
  "createLanguagePackWatcher",
]
`;
